### ISSUE TITLE
Examples not functioning

### ISSUE THREAD

**Issue by aultt** [VirtualEngine/Lability#172]
Working with both example TestLabGuide and demo from here -->https://github.com/iainbrighton/PSDEVOPS yield vm booting for an hour then shutting down.  Upon starting the vm back up windows does boot however the DSC configuration did not apply.  Looking through the logs the below error is found:
Job {AD35E038-B64D-11E6-80B6-00155D32F0EE} : 
MIResult: 6
Error Message: The PowerShell DSC resource MSFT_xIPAddress from module <xNetworking,2.12.0.0> does not exist at the PowerShell module path nor is it registered as a WMI DSC resource.
Message ID: DscResourceNotFound
Error Category: 7
Error Code: 6
Error Type: MI
Looking through the DSC resources all resources which were configured in the examples were indeed copied to the VM however resource.  I'm guessing I'm missing something environmentally anyone else see this issue?  Whats the expected time if SysPrep images exist to spin up a new lab with one DC and App Server?

**Comment by iainbrighton** [comment]
Hi @aultt - I looks like the .MOF files were compiled using a different version of the xNetworking module on the host, than the version copied in to the VM. If you run `Get-DscResource` on in the VM, I bet it's using a newer version?

It might be a good idea to update the examples to use the `RequiredVersion` hashtable key (rather than `MinimumVersion` ) to match the version of DSC resource modules installed on the host.

**Comment by andlo** [comment]
@iainbrighton I am looking into same problem after I did hack around my system.
Earlier I just commented out minimumsversion, as I then thaught that i would use the version om my host. So should i insert RequeredVersion = same as on host or what do you mean?

**Comment by iainbrighton** [comment]
@andlo @aultt There two things going on here:

* Lability uses the metadata in the .PSD1 file to determine what DSC resource modules to inject into the VM at creation time:
  * The modules are cached (by default) in the `%ProgramData%\Lability` folder.
  * The `MinimumVersion` check is performed against the modules cached in this directory.
  * If the minimum version isn't found the __latest__ version is downloaded.
  * What is installed/registered on the host is ignored; this permits us to support different versions of modules/resources on a v4 host.
* ___Lability does not check what version of resources are used in the MOF files___
  * The MOF files are copied in verbatim.
  * The DSC resource modules installed/found in `%ProgamFiles%\WindowsPowerShell\Modules` are used by PowerShell when creating the MOF files
  * There's no way to control what versions of modules will be used by the DSC engine when compiling the MOFs in the .PSD1 configuration file.

If there is any discrepancy between the versions copied/injected in to the VM and the module/resource versions defined in the resulting MOF files, you will have problems.

If you want Lability to use the DSC resource versions/modules installed on the host, __remove the `NonNodeData\Lability\DSCResource` key completely.__ Lability will then just copy all the DSC resources installed on the host, into the VMs.

It seems to me that the majority of people use MOF files compiled on the host and not precompiled MOF files. _Perhaps we should add a configuration switch to control whether to use the local host's DSC resource/modules?_

**Comment by iainbrighton** [comment]
Looping in @csandfeld, @kilasuit and @ryanspletzer for some additional feedback.. Thoughts on defaulting to the old/legacy behaviour?

**Comment by ryanspletzer** [comment]
**tl;dr**: I think it works great as it is right now and the only possible improvement could be telling Lability where to look for modules, but since multiple module versions can live side by side, I think `%ProgramFiles%\WindowsPowerShell\Modules` is still the most logical place to do that, so a switch to tell Lability to use the hosts' modules makes sense, but all the same rules re: `MinimumVersion`, `RequiredVersion` etc. should apply since the host *may* have multiple module versions installed. I think this really only solves an offline only scenario, though. A more judicious use of explicit `RequiredVersion` tags would solve most of these problems.

Hey @iainbrighton, so my typical usage pattern involves compiling .mof's on a Windows 10 host (latest build), and before I even get to the Lability parts of starting the lab configuration, etc., I have to have the modules installed in my Modules folder on the host in order to compile that .mof.

As a result of that, whenever I've been doing examples like Test Lab Guide (which I've used extensively as a jumping off point to further configurations), I'm using the latest module versions available via Install-Module. (By the way, TLG has been working great lately with newer modules.) It just so happened that Lability's downloading of the latest versions of the modules into the cache matched what I was doing on my host, by chance, so I had no issues. (I hadn't touched the `MinimumVersion` tags.)

![image](https://cloud.githubusercontent.com/assets/1481108/22614808/d309695a-ea55-11e6-88e8-388b454b78df.png)

That being said, I personally like the notion of explicitly calling out whatever resources / modules / dependencies in general and versions of them you need (like your packages.config in an ASP.NET app). You're not telling DSC what modules you need, you're telling Lability what modules you expect to be there via injection. And I like that the `MinimumVersion` and `RequiredVersion` tags work exactly like you would expect with the PackageManagement module. Maybe someone else's authoring box is separate from their Hyper-V host and they're pushing to it -- the configuration should still work in that scenario and Lability should attempt a "package restore" (to pull another analogy to ASP.NET / NuGet land).*

*The only caveat I can think of, as I was typing that, is if your Hyper-V host you're pushing to _doesn't have internet access_ for some reason. Then the modules need to be on the box (or on a share?) somewhere. We deal with this in ASP.NET land if we can't do NuGet package restores during build, we explicitly check in our packages alongside the source (yuck, but you have to deal sometimes), and we also can tell NuGet where to look for packages -- it has a a place to look by convention ( /Packages next to .sln or .csproj ) but it also can be told where to look for them.

So if Lability is doing this by convention:

> If you want Lability to use the DSC resource versions/modules installed on the host, **remove the `NonNodeData\Lability\DSCResource` key** completely. Lability will then just copy all the DSC resources installed on the host, into the VMs.

I think that's good. (Sidenote: does it copy all versions of all modules into the VM?)

If `NonNodeData\Lability\DSCResource` *is* present, it should attempt a restore based on info in those version tags. I think the only possible way you can improve upon that is you could also perhaps explicitly *tell* Lability where to look for modules (locally or on a share path) if you have a specific set you want copied in. I'm struggling to think of any case beyond that offline / no internet scenario where that even makes sense -- these things should be like rpm's, they're either available through a feed or you push them onto the box itself, especially since multiple module versions can live side by side now -- which brings us full circle, as most logical place to put modules is still `%ProgramFiles%\WindowsPowerShell\Modules` where multiple module versions can live side by side. So having that switch might make sense, but it's a very narrow use case I think.

**Comment by csandfeld** [comment]
@iainbrighton: Thanks for asking
@ryanspletzer : That was a really thorough post – thanks for sharing.

To be honest I haven't been doing much with Lability for quite a while - except (as luck would have it) I used yesterday to set up a small DFS test environment (worked a treat by the way). So I have actually had limited exposure to the “new” method with module cache etc. I do think however that whatever could be done to ensure successful lab deployment, would be worth doing.

Maybe Start-LabConfiguration should parse the Mof files, look for module versions used, and throw an error if those modules are not available on the host (searching in $PSModulePath folders and in Lability cache). Then we would also be able to ensure, that the right module versions are injected in the VM.

I have not tested any of this, but found the following when googling:
•	https://powershell.org/forums/topic/parse-mof-with-powershell/
•	https://github.com/mikeclayton/MofParser

The second link is C# library, and I think it would be better to stick with pure PowerShell code, but it shows that it is indeed possible to extract the information from a Mof.

I don’t know how much effort it would take to implement this, but I do think it would be a very nice feature.

**Comment by iainbrighton** [comment]
@ryanspletzer @csandfeld Thanks for the feedback! Where to start?!

The .mof mismatch is a mistake I see people making over and over. In essence it comes down to people using the `MinimumVersion` in the `DSCResource` declaration and not `RequiredVersion`. The examples don't help in this regard.

I am keen to avoid Lability doing things outside of its initial purpose of taking an existing DSC configuration and .mofs and creating VMs from them. However, adding a check that warns when modules defined in a .mof aren't on a system might help a little?

I have [some rudimentary code](https://gist.github.com/iainbrighton/9d3dd03630225ee44126769c5d9c50a9) that can parse a .mof file and return the module names/versions. Any chance you can check if this works against some of your .mof files correctly?

**Comment by iainbrighton** [comment]
@ryanspletzer If you were to check the compiled .mofs and zipped PS modules into version control, you could always run `Set-LabHostDefault -ModulePath <RepositoryRoot>\Modules` on the build server. Then Lability could find the exact module versions without internet access.. I think!

**Comment by csandfeld** [comment]
@iainbrighton I ran that Get-MofModule function against a bunch of my mofs, and at first glance it seem to extract the expected module and version info from each of them 👍 

Once we extract the module versions from the mofs, it should be pretty straight forward to ensure the right version of a module is injected, or throw an error that the module is not available. The more we can do to ensure a successful deployment, the better.

Personaly I don't see that as straying from Labilitys original purpose, but as it is based on your vision you know best :-)

**Comment by samirbazan** [comment]
I receive Error for SubnetMask using new modules of xNetworking:

ModuleType Version    Name
---------- -------    ----
Script     0.10.1     Lability

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Function        Convert-CIDRToSubhetMask                           3.2.0.0    xnetworking
Function        Get-LocalizedData                                  3.2.0.0    xnetworking
Function        Get-xNetworkAdapterName                            3.2.0.0    xnetworking
Function        New-InvalidArgumentException                       3.2.0.0    xnetworking
Function        New-InvalidOperationException                      3.2.0.0    xnetworking
Function        Set-xNetworkAdapterName                            3.2.0.0    xnetworking
Function        Test-IsNanoServer                                  3.2.0.0    xnetworking
Function        Test-xNetworkAdapterName                           3.2.0.0    xnetworking

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Function        Publish-DSCModuleAndMof                            6.0.0.0    xPSDesiredStateConfiguration
Function        Publish-ModuleToPullServer                         6.0.0.0    xPSDesiredStateConfiguration
Function        Publish-MOFToPullServer                            6.0.0.0    xPSDesiredStateConfiguration

The error need prefixlength instead of SubnetMask

At E:\Lability\Configurations\NanoExample.ps1:25 char:17
+                 SubnetMask     = $node.SubnetMask;
+                 ~~~~~~~~~~
The member 'SubnetMask' is not valid. Valid members are
'AddressFamily', 'DependsOn', 'InterfaceAlias', 'IPAddress', 'PrefixLength', 'PsDscRunAsCredential'.
    + CategoryInfo          : ParserError: (:) [], ParseException
    + FullyQualifiedErrorId : InvalidInstanceProperty

:S

**Comment by iainbrighton** [comment]
Hi @samirbazan yeah - that's unfortunate. In the [xNetworking v3.0.0.0](https://github.com/PowerShell/xNetworking/blob/dev/CHANGELOG.md#3000) release, the MSFT_xIPAddress:SubnetMask parameter was renamed to PrefixLength. This obviously breaks the bundled examples 😬.

I'll get the examples updated with the latest DSC resource modules. I don't think there are any other breaking changes?

**Comment by moodzuli** [comment]
Running on powershell x86 seems to course the problem on a 64bit machine
